torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -269,14 +260,33 @@ void ProcessGroupAgent::sync() {
 }
 
 void ProcessGroupAgent::start() {
+  rpcRunning_.store(true);
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
   futureTimeoutThread_ =
       std::thread(&ProcessGroupAgent::pollTimedOutRPCs, this);
 }
 
+void ProcessGroupAgent::shutdown() {
+  if (!rpcRunning_.exchange(false)) {
+    return;
+  }
+  LOG(INFO) << "Stopping ProcessGroupAgent.";
+  {
+    std::unique_lock<std::mutex> lock(recvWorkMutex_);
+    if (recvWork_) {
+      recvWork_->abort();
+    }
+  }
+  listenerThread_.join();
+  futureTimeoutCV_.notify_one();
+  futureTimeoutThread_.join();
+  // PythonRpcHandler::getInstance().cleanup();
```

Comment:
Any reason for doing cleanup in `join` instead of here?

Reply:
This wasn't right, I moved it to `shutdown()`.